### PR TITLE
HW2

### PR DESCRIPTION
**Pull request by Hannabolotova**
Create functions that find pop, top, min and add element in array

### CODE REVIEW

File: Desktop/CocoaSchool/.gitignore [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+*.xcodeproj/*
+!*.xcodeproj/project.pbxproj
+!*.xcodeproj/xcshareddata/
+!*.xcworkspace/contents.xcworkspacedata
+/*.gcno
+
+### Xcode Patch ###
+**/xcshareddata/WorkspaceSettings.xcsettings
+
+# End of https://www.gitignore.io/api/xcode,carthage,cocoapods,objective-c
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/.gitignore]: Please don't remove the end of line in the end of file.

File: Desktop/CocoaSchool/HW2.playground/Contents.swift [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    
+    override func top() -> Int? {
+        guard let topElement = array.last else {
+            return nil
+        }
+        return topElement
+    }
+    
+    func minimumElement() -> Int? {
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/HW2.playground/Contents.swift]: Could you please evaluate the function `minimumElement()` in big-O notation? Leave a comment here.

File: Desktop/CocoaSchool/HW2.playground/Contents.swift [hunk 3]
```diff
@@ -0,0 +1,72 @@
+class Stack{
+    fileprivate var array: [Int] = []
+    
+    func push(element: Int) {
+        array.append(element)
+    }
+    
+    func pop() -> Int? {
+        if array.isEmpty {
+            print("Array is empty")
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/HW2.playground/Contents.swift]: Please don't mix the logic (the behaviour of the stack) and the code that interacts to UI (`print` function here).
What if you need to print the information to UILabel, not to the console? That will require you to change the function that should be independent from UI.

Does that make sense?

File: Desktop/CocoaSchool/HW2.playground/Contents.swift (hunk 2 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/HW2.playground/Contents.swift]: O(n).  corrected to O(1) in the new commit.

File: Desktop/CocoaSchool/HW2.playground/Contents.swift (hunk 3 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/HW2.playground/Contents.swift]: corrected.

File: Desktop/CocoaSchool/.gitignore [hunk 4]
```diff
@@ -0,0 +1,103 @@
+
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/.gitignore]:
```suggestion
```

File: Desktop/CocoaSchool/HW2.playground/Contents.swift [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    
+    func top() -> Int? {
+        guard let topElement = array.last else {
+            return nil
+        }
+        return topElement
+    }
+}
+
+class StackStatistics: Stack{
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/HW2.playground/Contents.swift]:
```suggestion
class StackStatistics: Stack {
```

File: Desktop/CocoaSchool/HW2.playground/Contents.swift [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+minElement.push(element: 5)
+minElement.top()
+minElement.minimumElement()
+minElement.array = []
+minElement.minimumElement()
+minElement.top()
+minElement.push(element: 8)
+minElement.pop()
+
+
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/HW2.playground/Contents.swift]: Remove blank lines:

```suggestion
```

File: Desktop/CocoaSchool/HW2.playground/Contents.swift [hunk 7]
```diff
@@ -0,0 +1,64 @@
+class Stack{
+    fileprivate var array: [Int] = []
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/HW2.playground/Contents.swift]: With your current implementation I can do the following:

```swift
let s = Stack()
s.push(element: 1)
s.array.append(2)
s.top()
```

After the only one `push` I expect to see the value `1` at the top of the stack, but `s.top()` returns `2`.

File: Desktop/CocoaSchool/HW2.playground/Contents.swift [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    func pop() -> Int? {
+        if array.isEmpty {
+            return nil
+        } else {
+           return array.removeLast()
+        }
+    }
+    
+    func top() -> Int? {
+        guard let topElement = array.last else {
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/HW2.playground/Contents.swift]: Please read the documentation on `Array`'s last property.

File: Desktop/CocoaSchool/HW2.playground/Contents.swift [hunk 9]
```diff
@@ -0,0 +1,64 @@
+class Stack{
+    fileprivate var array: [Int] = []
+    
+    func push(element: Int) {
+        array.append(element)
+    }
+    
+    func pop() -> Int? {
+        if array.isEmpty {
```
Comment (dersim-davaod) [inline Desktop/CocoaSchool/HW2.playground/Contents.swift]: Why you did not use `popLast()` as below?

File: Desktop/CocoaSchool/HW2.playground/Contents.swift (hunk 7 above)
Comment (dersim-davaod) [reply Desktop/CocoaSchool/HW2.playground/Contents.swift]: The code violates the encapsulation.

As a consumer of the interface I don't need to know how you store the stack items. Say, you should be able to replace an array with a linked list, but the interface should remain the same. My code that calls `push()`/`pop()` functions won't be changed.

File: Desktop/CocoaSchool/.gitignore (hunk 1 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/.gitignore]: fixed.

File: Desktop/CocoaSchool/.gitignore (hunk 4 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/.gitignore]: done

File: Desktop/CocoaSchool/HW2.playground/Contents.swift (hunk 5 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/HW2.playground/Contents.swift]: corrected

File: Desktop/CocoaSchool/HW2.playground/Contents.swift (hunk 6 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/HW2.playground/Contents.swift]: corrected.

File: Desktop/CocoaSchool/HW2.playground/Contents.swift (hunk 7 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/HW2.playground/Contents.swift]: fixed.

File: Desktop/CocoaSchool/HW2.playground/Contents.swift (hunk 8 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/HW2.playground/Contents.swift]: thanks! corrected.

File: Desktop/CocoaSchool/HW2.playground/Contents.swift (hunk 9 above)
Comment (Hannabolotova) [reply Desktop/CocoaSchool/HW2.playground/Contents.swift]: corrected.

File: CocoaSchool/HW2.playground/Contents.swift [hunk 10]
```diff
@@ -0,0 +1,61 @@
+class Stack {
+    private var array: [Int] = []
```
Comment (dersim-davaod) [inline CocoaSchool/HW2.playground/Contents.swift]: nit: I suggest to give the more clear name.

File: CocoaSchool/HW2.playground/Contents.swift [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        return array.popLast()
+    }
+    
+    func top() -> Int? {
+        return array.last
+    }
+}
+
+class StackStatistics: Stack {
+    private var array: [Int] = []
```
Comment (dersim-davaod) [inline CocoaSchool/HW2.playground/Contents.swift]: I suggest to give more clear name.

**Pull request merged**